Enhance Qdrant Semantic Cache with flexible embedding dimensions and distance metrics

krishvsoni opened the pull request:
## Title

Implement flexible embedding dimensions for Qdrant Semantic Cache

## Relevant issues

Fixes ##9377

## Pre-Submission checklist

**Please complete all items before asking a LiteLLM maintainer to review your PR**

- [x] I have Added testing in the [`tests/litellm/`](https://github.com/BerriAI/litellm/tree/main/tests/litellm) directory, **Adding at least 1 test is a hard requirement** - [see details](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] I have added a screenshot of my new test passing locally 
- [x] My PR passes all unit tests on [`make test-unit`](https://docs.litellm.ai/docs/extras/contributing_code)
- [x] My PR's scope is as isolated as possible, it only solves 1 specific problem

## Type

🆕 New Feature

## Changes

- Added support for flexible embedding model dimensionality in LiteLLM's Qdrant Semantic Cache
- Implemented auto-detection of embedding dimensions by making a test embedding call during initialization
- Added manual configuration options for `qdrant_semantic_cache_embedding_dimensions` parameter
- Added support for configurable distance metrics (`Cosine`, `Dot`, `Euclid`) via `qdrant_distance_metric` parameter
- Updated Cache and QdrantSemanticCache classes to handle flexible dimensions and distance metrics
- Maintained full backward compatibility - all existing configurations continue to work without changes
- Added comprehensive test suite in `tests/test_litellm/caching/test_qdrant_flexible_dimensions.py`
- Created example script `examples/qdrant_flexible_dimensions_example.py` demonstrating usage
- Updated documentation in proxy configuration and caching guides to reflect new parameters
- Enables cost reduction by supporting cheaper embedding models (e.g., BGE Base 768d vs OpenAI 1536d)
- Allows access to higher-quality models with different dimensions (e.g., Stella 1024d, text-embedding-3-large 3072d)
- Supports any custom embedding model with dimensions from 64 to 8192

krishvsoni committed:
Enhance Qdrant Semantic Cache with flexible embedding dimensions and distance metrics

- Added support for specifying embedding dimensions and distance metrics in Cache and QdrantSemanticCache classes.
- Implemented auto-detection of embedding dimensions during initialization.
- Updated documentation to reflect new parameters and usage examples.
- Added demo and test scripts to showcase flexible embedding dimensions functionality.

krrishdholakia commented on litellm/caching/qdrant_semantic_cache.py:
```diff
@@ -89,6 +93,10 @@ def __init__(  # noqa: PLR0915
         )
 
+        # Auto-detect embedding dimensions if not provided
+        if self.embedding_dimensions is None:
```
making a real LLM API call on init can be dangerous 

we should try to have this happen outside init, and be opt in + cached where possible

krishvsoni committed:
Refactor embedding dimensions handling to lazy detection in QdrantSemanticCache

krishvsoni committed:
fix(tests): Add missing groq/llama-3.3-70b-versatile test case with correct function calling support

krishvsoni committed:
Merge branch 'main' into feature/qdrant-flexible-embedding-dimensions

krishvsoni committed:
chore(tests): Increase timeout and enhance test execution with performance monitoring

krishvsoni committed:
fix: optimize GitHub Actions timeout - increase timeout to 45min, improve test parallelism, add performance monitoring

- Increase timeout from 25min to 45min in test-litellm.yml
- Optimize test execution with 6 parallel workers (was 4)
- Add test filtering for fast/slow tests
- Add pytest-timeout to prevent hanging tests
- Add performance monitoring script
- Configure pytest.ini for better test management

Addresses: GitHub Actions job timeout issue in production

krishvsoni committed:
test: trigger workflow to verify timeout fixes

krishvsoni committed:
feat: add support for function calling and tool choice in chat mode

krishvsoni committed:
test: enhance test_invalid_purpose by mocking user_api_key_auth dependency

krishvsoni committed:
test: enhance test_invalid_purpose to prevent MagicMock interference and ensure proper mocking of dependencies

krishvsoni committed:
fix

krishvsoni committed:
Merge main into feature/qdrant-flexible-embedding-dimensions and resolve conflicts

krishvsoni committed:
fix: resolve ruff linting errors - remove unused import and suppress PLR0915

krishvsoni committed:
fix: remove unused variable assignments in guardrail handlers

krishvsoni committed:
feat: add request builders for various models in AmazonInvokeConfig and update tests for guardrail handlers

krishvsoni committed:
Merge branch 'main' into feature/qdrant-flexible-embedding-dimensions
